syntax/tags_test.go

```diff
@@ -13,3 +14,37 @@ func TestTagParsing(t *testing.T) {
 		t.Fatalf("Parsed fields incorrectly")
 	}
 }
+
+func TestTagValidity(t *testing.T) {
+	sliceTags := parseTag("head=2")
+	uintTags := parseTag("varint")
+	ptrTags := parseTag("optional")
+
+	sliceType := reflect.TypeOf([]byte{})
+	uintType := reflect.TypeOf(uint8(0))
+	ptrType := reflect.TypeOf(new(uint8))
+
+	if !tagsValidForType(sliceTags, sliceType) {
+		t.Fatalf("Rejected valid tags for slice")
+	}
+
+	if !tagsValidForType(uintTags, uintType) {
+		t.Fatalf("Rejected valid tags for uint")
+	}
+
+	if !tagsValidForType(ptrTags, ptrType) {
+		t.Fatalf("Rejected valid tags for ptr")
+	}
+
+	if tagsValidForType(uintTags, sliceType) {
+		t.Fatalf("Accepted invalid tags for slice: %v", uintTags)
+	}
+
+	if tagsValidForType(ptrTags, uintType) {
+		t.Fatalf("Accepted invalid tags for uint: %v", ptrTags)
+	}
+
+	if tagsValidForType(sliceTags, ptrType) {
+		t.Fatalf("Accepted invalid tags for ptr: %v", sliceTags)
+	}
```

Comment:
Don't you need 9 tests here?

Reply:
Even 9 tests wouldn't be sufficient, since there are other types :)  It seems enough to verify failure in at least one case.